### ISSUE TITLE
Add if! (if-bang) keyword to computation expressions

### ISSUE THREAD

**Issue by jwosty** [fsharp/fslang-suggestions#863]
# Title of Suggestion

I propose we extend computation expressions with a new keyword, `if!` (if-bang), that functions a ``let!` (let-bang) combined with an if statement. For example, instead of writing:

```
let! shouldDestroyWorld = confirmAsync ("Solve world hunger?")
if shouldDestroyWorld then solveWorldHunger () else dont ()
``` 

or even:

```
match! confirmAsync("Solve world hunger?") with
| true -> solveWorldHunger ()
| false -> dont ()
```

We should be able to do:

```
if! confirmAsync("Solve world hunger?")
then solveWorldHunger ()
else dont ()
```

## Pros and Cons 

Pros: Aids code readability in specialized cases.

Cons: It's work, and it increases complexity. (whether or not this is worth the effort for potentially such small return is up to debate / whoever would do it, but I figure it's worth filing [probably low priority :) ] since I am not aware of a prior suggestion -- would be brownie points for the F# language)

## Extra information

S

Related suggestions: (put links to related suggestions here)

#572

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [ x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [ x ] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [ x ] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [ x ] This is not a breaking change to the F# language design
* [ x ] I or my company would be willing to help implement and/or test this

**Comment by dsyme** [comment]
This seems reasonable

**Comment by nikonthethird** [comment]
Should that include `elif!` as well?

```fsharp
let x = async {
    if! asyncBool then
        do! task1
    elif! anotherAsyncBool then
        do! task2
    else
        do! task3
}
```

**Comment by yatli** [comment]
@nikonthethird probably yes, and pretty effortless, since `elif!` = `else if!`

**Comment by realvictorprm** [comment]
Good catch, that's really quite reasonable! :3

**Comment by mvkara** [comment]
It would be good to consider other typical branching constructs as well; in particular the "while" keyword. A good use case would be using ADO.NET; not sure if this is already supported. Normally you would need a recursive function to do this - would be much easier with .

```fsharp
asyncSeq {
    // https://docs.microsoft.com/en-us/dotnet/api/system.data.common.dbdatareader.readasync?view=netcore-3.1
    while! reader.MoveNextAsync() do
        yield {| Column1 = reader.GetInt32(0); Column2 = reader.GetInt32(1) |}
}
```

**Comment by Tarmil** [comment]
@mvkara There's a major difference between `while!` (and `for!` I guess) on one hand, and `match!` and `if!` on the other.

* `match!` and `if!` are just syntactic sugar: they're equivalent to `let!` followed by the corresponding keyword. So they don't require adding anything to the CE, as soon as they're added to the language they're usable with any CE.

* `while!`'s condition is not evaluated just once, so it's not so simple. It would need to be implemented explicitly by a CE (presumably as a method `WhileFrom`).

So I think it's a good idea but probably belongs in a separate suggestion.

**Comment by yatli** [comment]
I agree with @Tarmil -- plus that it's likely these constructs should be designed to incorporate with `IAsyncEnumerable`

**Comment by Lanayx** [comment]
Wanted to add one more comment for discussion, not directly related to the topic, but not sure where to put it - it would be nice to support bang operations inside **if**  without overhead of raising 0 to async:

```fsharp
async {
    let result = 
        if condition then
            let! x = doSmthAsyncReturningInt()
            x
        else
            0
    return result
}
```
this can be easily done in C#, but not F#:
```csharp
var result = condition ? await doSmthAsyncReturningInt() : 0;
```

**Comment by nikonthethird** [comment]
It is actually possible to implement a `while!` today by adjusting the `async` expression builder. You just have to add another overload for `while` that accepts an asynchronous guard:

```fsharp
type AsyncBuilder with
    member this.While (guard, body) =
        let guard = guard ()
        let rec loop = function
        | false -> this.Zero ()
        | _ -> this.Combine (body, this.Bind (guard, loop))
        this.Bind (guard, loop)
    

Async.RunSynchronously (async {
    let mutable count = 5
    while async { return count > 0 } do
        do printfn "Counting down: %d" count
        do count <- count - 1
})
```

**Comment by Tarmil** [comment]
@Lanayx The problem in your code isn't the `if` actually; it's that you're putting a `let!` inside the definition part of a `let`, which means it's not part of the CE anymore. See for example:

```fsharp
// This does NOT work, even without an if, because there's a let! inside a let:
async {
    let sum =
        let! x = someAsyncCode()
        x + y
    return sum * 2
}

// This DOES work, even with the if:
async {
    if someCondition() then
        let! x = someAsyncCode()
        return x + y
    else
        return y * 2
}
```

**Comment by Lanayx** [comment]
@Tarmil Thanks for the explanation! However as you can see, the problem is not in my code, but in this nested limitation - C# doesn't have such limitation and it would be nice to not have in F# either.

**Comment by yatli** [comment]
@Lanayx I think I understand the problem. But having different types on if-else branches feel wrong to me.

**Comment by Lanayx** [comment]
@yatli this is very helpful for caching scenarios - if local cache exists return cached value, otherwise make async call. To go over this limitation I had to [use mutable variable](https://github.com/fsharplang-ru/pulsar-client-dotnet/blob/schema/src/Pulsar.Client/Internal/ConsumerImpl.fs#L643-L649)

**Comment by yatli** [comment]
Hmm. Something like a new function `Async.FromCompleted<'T>`?

**Comment by Lanayx** [comment]
@yatli yes, if this cast is free :) With return it is an additional allocation and method call, and when it's on the hot path (like in my case), this should be avoided.

**Comment by sideeffffect** [comment]
could this be done, in a short-circuiting way?
```fsharp
let x = async {
    if! asyncBool and asyncBool2 then
        do! task1
    elif! anotherAsyncBool or asyncBool3 then
        do! task2
    else
        do! task3
}
```

**Comment by yatli** [comment]
@Lanayx I don't think it's optimized that way... More like a syntax sugar that wraps 0 into the completed task.

```C#
#LINQPad optimize+

async void Main()
{
	await compute(3);
}

// You can define other methods, fields, classes and namespaces here

async System.Threading.Tasks.Task<int> realAsync() {
	await System.Threading.Tasks.Task.Delay(100);
	return 0;
}

async System.Threading.Tasks.Task<int> compute(int x) => x > 2 ? await realAsync() : 0;
```

Turns to:
```MSIL
realAsync:
IL_0000:  ldloca.s    00 
IL_0002:  call        System.Runtime.CompilerServices.AsyncTaskMethodBuilder<System.Int32>.Create
IL_0007:  stfld       UserQuery+<realAsync>d__5.<>t__builder
IL_000C:  ldloca.s    00 
IL_000E:  ldc.i4.m1   
IL_000F:  stfld       UserQuery+<realAsync>d__5.<>1__state
IL_0014:  ldloca.s    00 
IL_0016:  ldflda      UserQuery+<realAsync>d__5.<>t__builder
IL_001B:  ldloca.s    00 
IL_001D:  call        System.Runtime.CompilerServices.AsyncTaskMethodBuilder<System.Int32>.Start<<realAsync>d__5>
IL_0022:  ldloca.s    00 
IL_0024:  ldflda      UserQuery+<realAsync>d__5.<>t__builder
IL_0029:  call        System.Runtime.CompilerServices.AsyncTaskMethodBuilder<System.Int32>.get_Task
IL_002E:  ret         

compute:
IL_0000:  ldloca.s    00 
IL_0002:  call        System.Runtime.CompilerServices.AsyncTaskMethodBuilder<System.Int32>.Create
IL_0007:  stfld       UserQuery+<compute>d__6.<>t__builder
IL_000C:  ldloca.s    00 
IL_000E:  ldarg.0     
IL_000F:  stfld       UserQuery+<compute>d__6.<>4__this
IL_0014:  ldloca.s    00 
IL_0016:  ldarg.1     
IL_0017:  stfld       UserQuery+<compute>d__6.x
IL_001C:  ldloca.s    00 
IL_001E:  ldc.i4.m1   
IL_001F:  stfld       UserQuery+<compute>d__6.<>1__state
IL_0024:  ldloca.s    00 
IL_0026:  ldflda      UserQuery+<compute>d__6.<>t__builder
IL_002B:  ldloca.s    00 
IL_002D:  call        System.Runtime.CompilerServices.AsyncTaskMethodBuilder<System.Int32>.Start<<compute>d__6>
IL_0032:  ldloca.s    00 
IL_0034:  ldflda      UserQuery+<compute>d__6.<>t__builder
IL_0039:  call        System.Runtime.CompilerServices.AsyncTaskMethodBuilder<System.Int32>.get_Task
IL_003E:  ret         
```

**Comment by Lanayx** [comment]
@yatli Right, for simple case it's obvious that result should be wrapped in completed task. However in case we have multiple async calls this can be avoided, here is a [sharplab](https://sharplab.io/#v2:EYLgdgpgLgZgHgHwAICYCMBYAUEgDAAiTQFYBubbJAZkJXwGF8BvbfNwmpADkIDYAeAJZgoAPnwAnCAEMANgEEAzgE8wAYwAUwqPjgBKZq3bGkATj4A6ACIRZ05RrS5ce8lmMmA7LrfGAvkZsgRyEPEgC2uIAsloiugYs7h5sAG7SEvjSaPgAvLr44p7eAPyE5lJySqqa+vgg+Li+yfhpGdJ0eVkF+Fw8pWaSMgoq6hpZBvWNwcatmTSddOKm5v3lQ1Wj7RMNTclI3tJUu/gBWH5AA==)

**Comment by yatli** [comment]
Thank you @Lanayx , I didn't know it could pull that off.
The C# version is obviously advantageous in your case, because the IL simply jumps from one unwrapped result to the next if the right conditions are met.

Would you mind opening another thread to continue the discussion? There are both syntax and compiler optimization issues, but out of scope of this suggestion.

**Comment by kurtschelfthout** [comment]
Consider allowing an elaboration to a new `If` method builder. Such a method would be somewhere between `Applicative` and `Monad` in power, and so can allow e.g. more efficient implementation, or conservative static analysis. See https://www.staff.ncl.ac.uk/andrey.mokhov/selective-functors.pdf

If F# automatically desugars it to `Bind` - as I think this suggestion is suggesting - then the implementor of the computation expression can't take that opportunity.

Perhaps a related suggestion then is to automatically desugar "lower power" computation expression operations in terms of explicitly defined "higher power" operations if the lower power operations are not available. E.g if `Bind` is available then `Map` can be generated in terms of `Bind`, and so on.

**Comment by Tarmil** [comment]
@kurtschelfthout It's a good point, although I would say it's a more general one that could apply to more than just `if!`. I know that I've wished for methods that would be able to transform simple `if` and `match` clauses in CEs. Specifically for Bolero's HTML CE. It needs information about the code structure to inform Blazor's diff algorithm, which currently is not available for `if` and `match`. So we have to avoid using them, and instead [use a function `cond` defined in Bolero](https://fsbolero.io/docs/HTML#conditional-elements).

I think we could do something like the following;

* Add a CE method `If()` that gets called when the CE contains an `if` clause.  
  It would take `(bool, M<'T>, M<'T>)` for `if..then..else`, and either `(bool, M<'T>)` or `bool, M<'T>, ce.Zero())` for `if..then`.  
  If absent, the current behavior applies.
* Add a syntax `if!` that becomes:
  * a call to `IfBind()`, if that exists, with similar arguments as `If()` except with `M<bool>`;
  * otherwise, a call to `Bind()` (ie the equivalent of `let!`) followed by the normal elaboration of `if` (ie either a call to `If` if it exists, or a normal conditional otherwise).
* For parity, do something similar for `match` and `match!`, with new methods `Match` and `MatchBind` that are called if they exist. Although I'm less sure what these methods should take as arguments since the structure of a `match` is more complex than that of an `if`.
